### DISCUSSION TITLE
REINFROCE or GRPO

### DISCUSSION THREAD

**Discussion by QingyuanWuNothing** [axon-rl/gem#71] [General]
Nice work! I’m a bit confused about the usage of REINFORCE and GRPO in the example.

1. The report states that multi-turn REINFORCE is used as the baseline (Algorithm 1). 
2. In the code, the default for reinforce_update is set to False ([link](https://github.com/sail-sg/oat/blob/eee3be4b0353153e11276d5814343bbf0c6febbc/oat/algorithms/ppo.py#L74)).
3. The provided example does not enable reinforce_update=True ([link](https://github.com/axon-rl/gem/tree/main/examples)).
4. In train.py, the learning_step of Learner is not modified ([link](https://github.com/axon-rl/gem/blob/main/examples/train_oat.py)). 

Did I miss something? Thanks!

**Comment by lkevinzc** [comment]
Hey @QingyuanWuNothing , thanks for your interest and your great questions!

To clarify, `reinforce_update` in oat is an old flag I used earlier to ablate the PPO's clipping operation, and it is not relevant to the multi-turn REINFORCE we presented in this work.

Instead, the multi-turn REINFORCE is achieved by:
1. considering transition level training samples (just as normal RL does) and compute turn-level returns via monte carlo ([link](https://github.com/axon-rl/gem/blob/main/examples/train_oat.py#L446)).
2. directly passing the computed returns as sample "reward" from actor to learner (a bit hacky, but in this way we can conveniently use this "reward" to reweigh policy gradient in learner, [link](https://github.com/axon-rl/gem/blob/main/examples/train_oat.py#L462))
3. on the learner side, we receive the reward at [here](https://github.com/sail-sg/oat/blob/main/oat/algorithms/ppo.py#L360), and processed [here](https://github.com/sail-sg/oat/blob/main/oat/algorithms/ppo.py#L448), then converted into advantage [here](https://github.com/sail-sg/oat/blob/main/oat/algorithms/ppo.py#L455).
4. In train_oat.py, we modify `compute_monte_carlo_advantages` ([code](https://github.com/axon-rl/gem/blob/main/examples/train_oat.py#L602)) to normalize the returns over the whole batch (as stated in Algo1).

**Comment by QingyuanWuNothing** [comment]
Thanks! I am really into your works.
1.  The MC return and advantage processing pipeline is cool and clear.
2.  So the loss calculation and `learning_step` in Learner used in this project is [here](https://github.com/sail-sg/oat/blob/main/oat/algorithms/ppo.py#L360)?  Please point to the correct one if I am wrong.

**Reply by lkevinzc** [reply]
Thank you very much for your kind words!

For the question: yep, because in `train_oat.py` we inherit `PPOLearner` from `oat` and customize it for the batch normalization.

I am really glad that you find our work useful. Please let me know any questions and I am happy to have further discussion!

**Reply by QingyuanWuNothing** [reply]
Thanks! No more questions here.
